utils/datasets.py

```diff
@@ -0,0 +1,282 @@
+            
+        Attributes:
+            None
+        
+        Methods:
+            get_MNIST() -> Dataset object:  
+                This method downloads the MNIST dataset in the ../data/MNIST folder if the dataset is not present.
+                Also returns the dataset in a PyTorch Dataset object which can be forwarded to a DataLoader as per the 
+                study's specification
+                
+            regression_fucntion() -> np.array:
+                This method takes in input data points and returns the output values points with a Gaussian Noise (0,0.02)
+                
+            get_regression() -> Train Dataset, Test Dataset:
+                This method returns a train and test dataset object for conducting the regression study. These dataset objects
+                can be forwarded to a DataLoader as per a person's requirement
+            
+            download_UCI() -> None:
+                (Under Development)
+                This method downloads the datasets for the reinforcement learning study (UCI_Mushroom Dataset). 
+                Note: This method doesn't return any Dataset object due to lack of understanding as to how the structure of the data should be
+                      With a better understanding of the study, this function can be enhanced to return a Dataset object as well.
+    """
+    def __init__(self):
+        pass
+    
+    def get_MNIST(self):
+        """
+            Method:
+                Returns 2 dataset objects (training and testing) with MNIST dataset in it.
+            Args:
+                None
+            Output:
+                (Dataset Object):   Returns the dataset object containing the training MNIST dataset
+                (Dataset Object):   Returns the dataset object containing the testing MNIST dataset
+        """
+        
+        # Generate the dataset object for MNIST
+        train_data = torchvision.datasets.MNIST('data/MNIST',transform=transforms.ToTensor(),download=True,train=True)
+        test_data = torchvision.datasets.MNIST('data/MNIST',train=False,transform=transforms.ToTensor())
+        
+        # Return the dataset object
+        return train_data,test_data
+    
+    def regression_function(self,x,sigma=0.02):
+        """
+            Method:
+                Returns the output values of the generating function with Gaussian Noise applied
+            Args:
+                x       (np.array): Array of Input Variables
+                sigma   (float)   : Variance of Gaussian Noise to be applied
+            Output:
+                (np.array)  : Numpy Array containing output values for each input values
+        """
+        
+        # Calculate the outputs with Gaussian Noise applied
+        y = x + 0.3 * np.sin(2*np.pi*(x+np.random.normal(0,sigma,x.shape))) + 0.3 * np.sin(4*np.pi*(x + np.random.normal(0,sigma,x.shape))) + np.random.normal(0,sigma,x.shape)
```

Comment:
well if you check there are default parameters used that are the same that were used in the paper. This epsilon was used everywhere and doesn't change.

Also the reason i did it was in case we wanna test it for more epsilons and check the performance. This way we can easily just change one value in the function and get the values.

Reply:
I think we are talking about different things, I am referring to the fact that in the paper it seems like it is the same sample everywhere, i.e. call np.random.normal once and store it in a separate variable, and then use it.

Reply:
Yes i get it and it is saved in the utils folder. the code checks for the files if they are present or not. Also there is a set seed so if there aren't the files present by any chance, it would generate the same numbers and store it

Reply:
so i'll re-run this code again and save it in npy files